# Phase 1
In Anti-Money Laundering (AML), especially in cryptocurrency, criminals don't just steal money and keep it still. They move it through complex webs of accounts to "clean" it (a process called layering). To catch them, we cannot look at transactions in a vacuum; we must look at the relationships between them.

This brings us to Graph Networks. In a graph:
- Nodes are entities (in our case, Bitcoin transactions).
- Edges are the connections between them (the flow of Bitcoin from one transaction to the next).
The goal of this phase is to built a model that catches illicit transactions not just based on the transaction's own features, but based on who it interacts with in the network.

In [8]:
import pandas as pd
from pathlib import Path

The Elliptic dataset is the world's largest labeled dataset of cryptocurrency transactions. It contains over 200,000 Bitcoin transactions.
- Some are labeled as "illicit" (tied to scams, malware, terrorist organizations, etc.).
- Some are labeled as "licit" (exchanges, wallet providers, miners).
- The rest are "unknown.

In [ ]:
import kaggle
kaggle datasets download ellipticco/elliptic-data-set

In [ ]:
#labels
# unknown, 2 (licit), and 1 (illicit).
df_classes = pd.read_csv(Path("elliptic_bitcoin_dataset") / "elliptic_txs_classes.csv")
# two columns of source and destination
df_edgelist = pd.read_csv(Path("elliptic_bitcoin_dataset") / "elliptic_txs_edgelist.csv")
# 165 anonymised mathematical features of every transaction, and a time step from 1 to 49 representing two week intervals. no column names for the features file, so we set header=None
df_features = pd.read_csv(Path("elliptic_bitcoin_dataset") / "elliptic_txs_features.csv", header=None)

print("Classes Overview")
print(df_classes.head())
print("\nClass distribution")
print(df_classes["class"].value_counts())


Classes Overview
        txId    class
0  230425980  unknown
1    5530458  unknown
2  232022460  unknown
3  232438397        2
4  230460314  unknown

Class distribution
class
unknown    157205
2           42019
1            4545
Name: count, dtype: int64


**Observation:** 
77% of our data (157,205 transactions) is "unknown". This reflects the harsh reality of AML: investigators simply don't have the time to investigate and label every single Bitcoin transaction. We only know the true labels for a small fraction of the network. This pushes us into the realm of Semi-Supervised Learning or Node Classification.

For now, to train our initial models, we will only use the transactions we know the answers to (the 1s and 2s). Later, we can use the graph structure to let the "bad" reputation of known illicit nodes bleed over into the "unknown" nodes they interact with.